Loop/View Models/SimpleBolusViewModel.swift

```diff
@@ -0,0 +1,332 @@
+        case saveWithoutBolusing
+        case saveAndDeliver
+        case enterBolus
+        case deliver
+    }
+    
+    var hasDataToSave: Bool {
+        return glucose != nil || carbs != nil
+    }
+    
+    var hasBolusEntryReadyToDeliver: Bool {
+        return bolus != nil
+    }
+    
+    var actionButtonAction: ActionButtonAction {
+        switch (hasDataToSave, hasBolusEntryReadyToDeliver) {
+        case (true, true): return .saveAndDeliver
+        case (true, false): return .saveWithoutBolusing
+        case (false, true): return .deliver
+        case (false, false): return .enterBolus
+        }
+    }
+    
+    var carbPlaceholder: String {
+        Self.carbAmountFormatter.string(from: 0.0)!
+    }
+
+    private let glucoseAmountFormatter: NumberFormatter
+    private let delegate: SimpleBolusViewModelDelegate
+    private let log = OSLog(category: "SimpleBolusViewModel")
+    
+    private lazy var bolusVolumeFormatter = QuantityFormatter(for: .internationalUnit())
+
+    var maximumBolusAmountString: String? {
+        return bolusVolumeFormatter.numberFormatter.string(from: delegate.maximumBolus) ?? String(delegate.maximumBolus)
+    }
+
+    init(delegate: SimpleBolusViewModelDelegate) {
+        self.delegate = delegate
+        let glucoseQuantityFormatter = QuantityFormatter()
+        glucoseQuantityFormatter.setPreferredNumberFormatter(for: delegate.preferredGlucoseUnit)
+        glucoseAmountFormatter = glucoseQuantityFormatter.numberFormatter
+        enteredBolusAmount = Self.doseAmountFormatter.string(from: 0.0)!
+        updateRecommendation()
+    }
+    
+    func updateRecommendation() {
+        if carbs != nil || glucose != nil {
+            recommendation = delegate.computeSimpleBolusRecommendation(mealCarbs: carbs, manualGlucose: glucose)?.doubleValue(for: .internationalUnit())
+        } else {
+            recommendation = nil
+        }
+    }
+    
+    func saveAndDeliver(completion: @escaping (Bool) -> Void) {
+        if let bolus = bolus {
+            guard bolus.doubleValue(for: .internationalUnit()) <= delegate.maximumBolus else {
+                presentAlert(.maxBolusExceeded)
+                completion(false)
+                return
+            }
+        }
+
+        if let glucose = glucose {
+            guard LoopConstants.validManualGlucoseEntryRange.contains(glucose) else {
+                presentAlert(.manualGlucoseEntryOutOfAcceptableRange)
+                completion(false)
+                return
+            }
+        }
+        
+        if let carbs = carbs {
+            guard carbs <= LoopConstants.maxCarbEntryQuantity else {
+                presentAlert(.carbEntrySizeTooLarge)
+                completion(false)
+                return
+            }
+        }
+        
+        let saveDate = Date()
+
+        // Authenticate the bolus before saving anything
+        func authenticateIfNeeded(_ completion: @escaping (Bool) -> Void) {
```

Comment:
Hm.  I haven't seen this in common usage: declaring functions inside functions?  I like how this makes the actual `saveAndDeliver` code below easy to understand, but is this really necessary, or could they just be declared `private` at class scope?  It honestly makes the code harder to read for me.

Reply:
So the scoping here is for variable access (to `saveDate`, for one), and for organization; it's a sign to readers that these methods are only used within this function, and not called elsewhere.

I understand that it's additional indentation and not very common, but are there other aspects that you think contribute to making it harder to parse?

Reply:
implicit variable capture is in fact one of the reasons it is harder to parse.  Consider pulling them out and passing the variables in.

Reply:
Thanks for sharing your input on this. I think it's a tradeoff; in the local scope case, it's easily verifiable that the same date is used by all the methods; splitting them out makes the user trace them (now there are x copies of save date, and more boilerplate to pass them around that could be sources of new errors). And I really like the scope containment with this approach. It's similar to having a separate class for just doing the save; the concerns around this are better localized, imo.

Reply:
Okay. I'll leave it up to you then.